### PR TITLE
Introduce AST, MQL parser, renderer

### PR DESCRIPTION
**Pull request by johnnyshields**

### CODE REVIEW

File: lib/active_document/ast/node/nin.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+module ActiveDocument
+  module AST
+    class Nin < Node
```
Comment (johnnyshields) [inline lib/active_document/ast/node/nin.rb:5]: Rename to `NotIn`

File: lib/active_document/ast/node/ne.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+module ActiveDocument
+  module AST
+    class Ne < Node
```
Comment (johnnyshields) [inline lib/active_document/ast/node/ne.rb:5]: rename to `NotEq`

File: lib/active_document/parser/mql.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module ActiveDocument
+  module Parser
+    module MQL
+      LIST_COMPARISON_OPERATORS = %i[eq gt gte in lt lte ne nin].freeze
+      LIST_LOGICAL_OPERATORS = %i[and nor not or].freeze
+      LIST_OPERATORS = (LIST_COMPARISON_OPERATORS + LIST_LOGICAL_OPERATORS).freeze
+
+      MAP_OPERATORS = LIST_OPERATORS.map {|op| [op, [ActiveDocument::AST, op.capitalize].join('::').constantize] }.to_h.freeze
```
Comment (johnnyshields) [inline lib/active_document/parser/mql.rb:10]: Please make this an explicit list like:

     MAP = {
        '$all' => All,
        '$bitsAllClear' => BitsAllClear,
        '$bitsAllSet' => BitsAllSet,
        '$bitsAnyClear' => BitsAnyClear,
        '$bitsAnySet' => BitsAnySet,
        '$elemMatch' => ElemMatch,
        '$eq' => Eq,
        '$exists' => Exists,
        '$gt' => Gt,
        '$gte' => Gte,
        '$in' => In,
        '$lt' => Lt,
        '$lte' => Lte,
        '$mod' => Mod,
        '$nin' => Nin,
        '$ne' => Ne,
        '$not' => Not,
        '$regex' => Regex,
        '$size' => Size,
        '$type' => Type
      }.freeze

File: lib/active_document/ast/node/and.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+module ActiveDocument
+  module AST
+    class And < Node
```
Comment (johnnyshields) [inline lib/active_document/ast/node/and.rb:5]: Introduce a `LogicalOperator` intermediate class

Make `And < LogicalOperator < Node`

File: lib/active_document/parser/mql.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+module ActiveDocument
+  module Parser
+    module MQL
+      LIST_COMPARISON_OPERATORS = %i[eq gt gte in lt lte ne nin].freeze
```
Comment (johnnyshields) [inline lib/active_document/parser/mql.rb:6]: call these FIELD_OPERATORS

File: lib/active_document/parser/mql.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      ValueNode = ActiveDocument::AST::Value
+      AndNode = ActiveDocument::AST::And
+
+      class << self
+
+        def parse(mql_hash)
+          conditions = []
+          mql_hash.each do |key, sub_hash|
+            sub_hash.each do |op, value|
+              conditions << node_klass(op).new([FieldNode.new([key]), ValueNode.new([value])])
```
Comment (johnnyshields) [inline lib/active_document/parser/mql.rb:22]: I think we don't need FieldNode and ValueNode classes. Instead just do:

```ruby
class Eq < FieldOperator
end

class FieldOperator < Node
  def initialize(field, value)
    @field = field
    @value = value
  end

   attr_reader :field, :value
end

class Or < LogicalOperator
end

class LogicalOperator < Node
  def initialize(conditions)
    @conditions = conditions #array
  end

   attr_reader :conditions
end
```

File: lib/active_document/parser/mql.rb [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        end
+
+        def do_parse(data, path = [])
+          case data
+          when Hash
+            if data.size == 1
+              do_parse(data.values[0], path + [data.keys[0]])
+            else
+              head = data.select.with_index { |_, i| i == 0 }
+              tail = data.reject.with_index { |_, i| i == 0 }
```
Comment (johnnyshields) [inline lib/active_document/parser/mql.rb:40]: can use `.partition` here

File: lib/active_document/parser/mql.rb [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+      OPERATORS = [FIELD_OPERATORS, LOGICAL_OPERATORS].inject(&:merge).freeze
+
+      class << self
+
+        def parse(mql_hash)
+          do_parse(mql_hash)
+        end
+
+        def do_parse(data, path = [])
```
Comment (johnnyshields) [inline lib/active_document/parser/mql.rb:33]: split this into parse_hash, parse_array, parse_numeric, etc.

File: lib/active_document/parser/mql.rb (hunk 7 above)
Comment (SavickyAnt) [reply lib/active_document/parser/mql.rb:40]: Like so?
```
h, t = {a:1, b:2, c:3}.partition.with_index { |_, i| i == 0 }
head, tail = h.to_h, t.to_h
```

File: spec/active_document/parser/mql_spec.rb [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        let(:criteria) { Person.where(age: { '$eq' => 0 }).where(age: { '$gt' => 1 }).where(age: { '$lte' => -1 }).where(age: { '$in' => [2, 1, 0, -1, -2] }) }
+
+        it { is_expected.to eq("(And:\n  (And:\n    (And:\n      (AnyIn: field=\"age\", value=[2, 1, 0, -1, -2]), \n      (Lte: field=\"age\", value=-1)), \n    (Gt: field=\"age\", value=1)), \n  (Eq: field=\"age\", value=0))") }
+      end
+
+      context 'with many different operators with repetitions' do
+        let(:criteria) { Person.where(age: { '$eq' => 0 }).where(age: { '$gt' => 1 }).where(age: { '$lte' => -1 }).where(age: { '$in' => [2, 1, 0, -1, -2] }).where(age: { '$eq' => -10 }).where(age: { '$gt' => 10 }) }
+
+        it {
+          is_expected.to eq("(And:\n  (And:\n    (And:\n      (And:\n        (And:\n          (Gt: field=\"age\", value=10), \n          (Eq: field=\"age\", value=-10)), \n        (AnyIn: field=\"age\", value=[2, 1, 0, -1, -2])), \n      (Lte: field=\"age\", value=-1)), \n    (Gt: field=\"age\", value=1)), \n  (Eq: field=\"age\", value=0))")
```
Comment (johnnyshields) [inline spec/active_document/parser/mql_spec.rb:89]: please define `==` operator on nodes, and then actually build the node tree.

Also should define the `class ()` operator so we can do `And(And(AnyIn(:age, [1,3,4]), Eq(:name, "Bob")))`

**Comment by johnnyshields** [comment]
Let's split out SimpleCov stuff to a separate PR. Fine to just run it locally for now

**Pull request merged**